src/main/java/com/zotov/edu/passportofficerestservice/controller/PersonApiController.java

```diff
@@ -0,0 +1,138 @@
+
+    private final PassportService passportService;
+
+    private final PassportControllerDtoConverter passportControllerDtoConverter;
+
+    @GetMapping
+    public Page<PersonResponse> getPersons(@RequestParam(required = false) Optional<String> passportNumber,
+                                           @PageableDefault(size = 100) Pageable pageable) {
+        Page<Person> foundPersons = passportNumber
+                .map(personsService::getPersonByPassportNumber)
+                .orElseGet(() -> personsService.getAllPersons(pageable));
+        return passportControllerDtoConverter.convertPersonToDto(foundPersons);
+    }
+
+    @PostMapping()
+    @ResponseStatus(HttpStatus.CREATED)
+    public PersonResponse createPerson(@Valid @RequestBody PersonRequest person) {
+        Person createdPerson = personsService.createPerson(person.getName(),
+                person.getBirthday(), person.getCountry());
+        return passportControllerDtoConverter.convertPersonToDto(createdPerson);
+    }
+
+    @GetMapping("/{id}")
+    public PersonResponse getPerson(@PathVariable String id) {
+        return passportControllerDtoConverter.convertPersonToDto(personsService.getPerson(id));
+    }
+
+    @PutMapping("/{id}")
+    public PersonResponse updatePerson(@PathVariable String id,
+                                       @Valid @RequestBody PersonRequest person) {
+        Person updatedPerson = personsService.updatePerson(
+                id, person.getName(), person.getBirthday(), person.getCountry());
+        return passportControllerDtoConverter.convertPersonToDto(updatedPerson);
+    }
+
+    @DeleteMapping("/{id}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void deletePerson(@PathVariable String id) {
+        personsService.deletePersonById(id);
+    }
+
+    @GetMapping("/{id}/passports")
+    public List<PassportResponse> getPassportsByGivenDatesRange(@PathVariable String id,
+                                                                @RequestParam(required = false, defaultValue = "ACTIVE")
+                                                                        PassportState state,
+                                                                @RequestParam(required = false)
+                                                                @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate minGivenDate,
+                                                                @RequestParam(required = false)
+                                                                @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate maxGivenDate) {
+        personsService.checkIfPersonExists(id);
+        List<Passport> foundPassports;
+        if (minGivenDate != null || maxGivenDate != null) {
+            foundPassports = passportService.getPassportsByGivenDateRange(id, state, minGivenDate, maxGivenDate);
+        } else {
+            foundPassports = passportService.getPassports(id, state);
```

Comment:
It seems over complicated to have several methods for a combination of nullable parameters. Better to have a single one that makes all the filtering. It's possible for both collections repository implementation and jdbc

Reply:
The methods have been replaced by the new method.